n2survey/read_from_xml.py

```diff
@@ -0,0 +1,268 @@
+                    "is_sub": False,
+                    "choices": {"I don't want to answer this question": 'A2', ...},
+                    "section_id": 13901,
+                    "question group": A1
+                },
+                {
+                    "name": "A4[other]",
+                    "question": "What is your year of birth? other",
+                    "description": "Please use the following format: 'xxxx' , e.g. 1989",
+                    "question type": "text",
+                    "is_other": True,
+                    "is_sub": False,
+                    "choices": {},
+                    "section_id": 13901,
+                    "question group": A4
+                },
+                {
+                    "name": "C5_SQ003",
+                    "question": "Do you get external financial support to cover...",
+                    "description": "",
+                    "question type": "multi choice",
+                    "is_other": False,
+                    "is_sub": False,
+                    "chioces": {"Parents": "Y"},
+                    "section_id": 13907 ,
+                    "question group": C5T
+                },
+                ...
+            }
+        }
+
+    Raises:
+        ValueError: Invalid XML file
+        ValueError: Provided file has unexpected structure
+    """
+    ans = dict()
+    try:
+        f = open(filepath, "r")
+    except OSError:
+        print(f"{filepath} is an invalid xml file")
+        sys.exit()
+    f = open(filepath, "r")
+
+    soup = BeautifulSoup(f, "xml")
+    sections = soup.find_all("section")
+    res_sections = list()
+    res_questions = list()
+
+    for section in sections:
+        section_id = section.attrs["id"]
+        section_title = section.sectionInfo.find("text").get_text()
+
+        for info in section.find_all("sectionInfo"):
+            section_text = info.find_all("text")
+            text_p = ""
+            for t in section_text:
+                text_p += _sectioninfo_text_parse(t.string)
+            section_info = text_p
+        res_sections.append(
+            {"id": section_id, "title": section_title, "info": section_info}
+        )
+        questions = section.find_all("question")
+
+        for question in questions:
+            question_content = _html_purify(question.find("text").string)
+            if question.find("directive") is not None:
+                question_descs = question.find("directive").find_all("text")
+                question_desc = _concat_question_text(question_descs)
+            else:
+                question_desc = ""
+            stored_questions = _store_questions(
+                question, section_id, question_content, question_desc
+            )
+            subquestion_dict = list()
+            if bool(question.subQuestion):
+                subquestion_dict = _integrate_subquestions(question, section_id)
+            res_questions += subquestion_dict + stored_questions
```

Comment:
This part is a little bit confusing. I would suggest adding one function that parses the question like 
```python
for question in questions:
    res_questions += _parse_question(question)
```

Then this function `_parse_question` returns all columns, i.e. for single choice + other it will be 2 items: "A1" and "A1other"; for a group of sub-questions each of them will be as a separate item.

Reply:
ok, thank you, it has been changed